File: src/VbNet.LanguageServer/Protocol/NamedPipeTransport.cs
// Named pipe transport implementation for LSP communication
// Primary transport - follows C# extension pattern (outputs pipe name as JSON to stdout)

using System.IO.Pipes;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace VbNet.LanguageServer.Protocol;

/// <summary>
/// LSP transport using named pipes (Windows) or Unix domain sockets (macOS/Linux).
/// This is the primary transport, matching the C# extension behavior.
///
/// Protocol:
/// 1. Server creates a named pipe with a unique name
/// 2. Server outputs {"pipeName":"..."} to stdout
/// 3. Client reads this and connects to the pipe
/// 4. Bidirectional LSP communication over the pipe
/// </summary>
public sealed class NamedPipeTransport : ITransport
{
    private readonly ILogger<NamedPipeTransport> _logger;
    private readonly string _pipeName;
    private readonly SemaphoreSlim _writeLock = new(1, 1);

    private NamedPipeServerStream? _pipeServer;
    private Stream? _readStream;
    private Stream? _writeStream;
    private bool _isConnected;

    private const string ContentLengthHeader = "Content-Length: ";

    public NamedPipeTransport(ILogger<NamedPipeTransport> logger, string? pipeName = null)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _pipeName = pipeName ?? GeneratePipeName();
    }

    /// <summary>
    /// Gets the pipe name for this transport.
    /// </summary>
    public string PipeName => _pipeName;

    /// <summary>
    /// Generates a unique pipe name appropriate for the current platform.
    /// </summary>
    private static string GeneratePipeName()
    {
        var uniqueId = Guid.NewGuid().ToString("N")[..8];

        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            // Windows named pipes don't need a path prefix in the API
            return $"vbnet-lsp-{uniqueId}";
        }
        else
        {
            // Unix domain sockets need a file path
            var tempDir = Path.GetTempPath();
            return Path.Combine(tempDir, $"vbnet-lsp-{uniqueId}.sock");
        }
    }

    public async Task StartAsync(CancellationToken cancellationToken = default)
    {
        _logger.LogDebug("Creating named pipe: {PipeName}", _pipeName);

        // Create the named pipe server
        _pipeServer = new NamedPipeServerStream(
            _pipeName,
            PipeDirection.InOut,
            maxNumberOfServerInstances: 1,
            PipeTransmissionMode.Byte,
            PipeOptions.Asynchronous);

        // IMPORTANT: Start the async wait for connection BEFORE outputting the pipe name.
        // This ensures the pipe is fully ready to accept connections when the client
        // reads the pipe name from stdout. The client will connect, which completes
        // the WaitForConnectionAsync task.
        //
        // Protocol synchronization:
        // 1. Server creates pipe and starts listening (WaitForConnectionAsync begins)
        // 2. Server outputs pipe name to stdout (signals readiness)
        // 3. Client reads pipe name and connects
        // 4. WaitForConnectionAsync completes
        _logger.LogDebug("Starting to listen on pipe: {PipeName}", _pipeName);
        var connectionTask = _pipeServer.WaitForConnectionAsync(cancellationToken);

        // Small delay to ensure the pipe is fully registered in the OS
        // before we signal readiness to the client
        await Task.Delay(50, cancellationToken);

        // NOW output the pipe name - we're ready to accept connections
        var pipeInfo = JsonSerializer.Serialize(new { pipeName = GetFullPipePath() });
        await Console.Out.WriteLineAsync(pipeInfo);
        await Console.Out.FlushAsync();
        _logger.LogInformation("Pipe ready, name output to stdout: {PipeInfo}", pipeInfo);

        // Wait for client to connect
        _logger.LogDebug("Waiting for client connection on pipe: {PipeName}", _pipeName);
        await connectionTask;

        _readStream = _pipeServer;
        _writeStream = _pipeServer;
        _isConnected = true;

        _logger.LogInformation("Client connected to named pipe");
    }

    /// <summary>
    /// Gets the full pipe path that clients should use to connect.
    /// </summary>
    private string GetFullPipePath()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            return $@"\\.\pipe\{_pipeName}";
        }
        else
        {
            // On Unix, the pipe name is already the full socket path
            return _pipeName;
        }
    }

    public async Task<string?> ReadMessageAsync(CancellationToken cancellationToken = default)
    {
        if (!_isConnected || _readStream == null)
        {
            throw new InvalidOperationException("Transport is not connected");
        }

        try
        {
            var contentLength = await ReadContentLengthAsync(cancellationToken);
            if (contentLength == null)
            {
                _logger.LogDebug("End of pipe stream reached");
                return null;
            }

            var buffer = new byte[contentLength.Value];
            var totalRead = 0;
            while (totalRead < contentLength.Value)
            {
                var bytesRead = await _readStream.ReadAsync(
                    buffer.AsMemory(totalRead, contentLength.Value - totalRead),
                    cancellationToken);

                if (bytesRead == 0)
                {
                    _logger.LogWarning("Unexpected end of stream while reading message body");
                    return null;
                }

                totalRead += bytesRead;
            }

            var message = Encoding.UTF8.GetString(buffer);
            _logger.LogTrace("Received message via pipe: {Message}", message);
            return message;
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error reading message from named pipe");
            throw;
        }
    }

    public async Task WriteMessageAsync(string message, CancellationToken cancellationToken = default)
    {
        if (!_isConnected || _writeStream == null)
        {
            throw new InvalidOperationException("Transport is not connected");
        }

        var messageBytes = Encoding.UTF8.GetBytes(message);
        var header = $"{ContentLengthHeader}{messageBytes.Length}\r\n\r\n";
        var headerBytes = Encoding.UTF8.GetBytes(header);

        await _writeLock.WaitAsync(cancellationToken);
        try
        {
            await _writeStream.WriteAsync(headerBytes, cancellationToken);
            await _writeStream.WriteAsync(messageBytes, cancellationToken);
            await _writeStream.FlushAsync(cancellationToken);
            _logger.LogTrace("Sent message via pipe: {Message}", message);
        }
        finally
        {
            _writeLock.Release();
        }
    }

    private async Task<int?> ReadContentLengthAsync(CancellationToken cancellationToken)
    {
        if (_readStream == null) return null;

        var headerBuilder = new StringBuilder();
        var buffer = new byte[1];

        while (true)
        {
            var bytesRead = await _readStream.ReadAsync(buffer, cancellationToken);
            if (bytesRead == 0)
            {
                return null;
            }

            headerBuilder.Append((char)buffer[0]);
            var headerText = headerBuilder.ToString();

            if (headerText.EndsWith("\r\n\r\n", StringComparison.Ordinal))
            {
                var lines = headerText.Split("\r\n", StringSplitOptions.RemoveEmptyEntries);
                foreach (var line in lines)
                {
                    if (line.StartsWith(ContentLengthHeader, StringComparison.OrdinalIgnoreCase))
                    {
                        var lengthStr = line.Substring(ContentLengthHeader.Length);
                        if (int.TryParse(lengthStr, out var length))
                        {
                            return length;
                        }
                        _logger.LogError("Invalid Content-Length value: {Value}", lengthStr);
                        throw new InvalidOperationException($"Invalid Content-Length: {lengthStr}");
                    }
                }

                _logger.LogError("No Content-Length header found in: {Headers}", headerText);
                throw new InvalidOperationException("Missing Content-Length header");
            }

            if (headerBuilder.Length > 8192)
            {
                throw new InvalidOperationException("Header too large");
            }
        }
    }

    public async ValueTask DisposeAsync()
    {
        _isConnected = false;
        _writeLock.Dispose();

        if (_pipeServer != null)
        {
            await _pipeServer.DisposeAsync();
        }

        // Clean up Unix socket file if it exists
        if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && File.Exists(_pipeName))
        {
            try
            {
                File.Delete(_pipeName);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to delete socket file: {Path}", _pipeName);
            }
        }
    }
}

File: src/VbNet.LanguageServer.Vb/Protocol/NamedPipeTransport.vb
' Named pipe transport implementation for LSP communication
' Primary transport - follows C# extension pattern (outputs pipe name as JSON to stdout)

Imports System.IO.Pipes
Imports System.Runtime.InteropServices
Imports System.Text
Imports System.Text.Json
Imports Microsoft.Extensions.Logging

Namespace Protocol

    ''' <summary>
    ''' LSP transport using named pipes (Windows) or Unix domain sockets (macOS/Linux).
    ''' This is the primary transport, matching the C# extension behavior.
    '''
    ''' Protocol:
    ''' 1. Server creates a named pipe with a unique name
    ''' 2. Server outputs {"pipeName":"..."} to stdout
    ''' 3. Client reads this and connects to the pipe
    ''' 4. Bidirectional LSP communication over the pipe
    ''' </summary>
    Public NotInheritable Class NamedPipeTransport
        Implements ITransport

        Private ReadOnly _logger As ILogger(Of NamedPipeTransport)
        Private ReadOnly _pipeName As String
        Private ReadOnly _writeLock As SemaphoreSlim = New SemaphoreSlim(1, 1)

        Private _pipeServer As NamedPipeServerStream
        Private _readStream As Stream
        Private _writeStream As Stream
        Private _isConnected As Boolean

        Private Const ContentLengthHeader As String = "Content-Length: "

        Public Sub New(logger As ILogger(Of NamedPipeTransport), Optional pipeName As String = Nothing)
            If logger Is Nothing Then
                Throw New ArgumentNullException(NameOf(logger))
            End If

            _logger = logger
            _pipeName = If(String.IsNullOrWhiteSpace(pipeName), GeneratePipeName(), pipeName)
        End Sub

        ''' <summary>
        ''' Gets the pipe name for this transport.
        ''' </summary>
        Public ReadOnly Property PipeName As String
            Get
                Return _pipeName
            End Get
        End Property

        ''' <summary>
        ''' Generates a unique pipe name appropriate for the current platform.
        ''' </summary>
        Private Shared Function GeneratePipeName() As String
            Dim uniqueId = Guid.NewGuid().ToString("N").Substring(0, 8)

            If RuntimeInformation.IsOSPlatform(OSPlatform.Windows) Then
                ' Windows named pipes don't need a path prefix in the API
                Return $"vbnet-lsp-{uniqueId}"
            End If

            ' Unix domain sockets need a file path
            Dim tempDir = Path.GetTempPath()
            Return Path.Combine(tempDir, $"vbnet-lsp-{uniqueId}.sock")
        End Function

        Public Async Function StartAsync(Optional cancellationToken As CancellationToken = Nothing) As Task Implements ITransport.StartAsync
            _logger.LogDebug("Creating named pipe: {PipeName}", _pipeName)

            _pipeServer = New NamedPipeServerStream(
                _pipeName,
                PipeDirection.InOut,
                maxNumberOfServerInstances:=1,
                PipeTransmissionMode.Byte,
                PipeOptions.Asynchronous)

            _logger.LogDebug("Starting to listen on pipe: {PipeName}", _pipeName)
            Dim connectionTask = _pipeServer.WaitForConnectionAsync(cancellationToken)

            Await Task.Delay(50, cancellationToken).ConfigureAwait(False)

            Dim pipeInfo = JsonSerializer.Serialize(New With {.pipeName = GetFullPipePath()})
            Await System.Console.Out.WriteLineAsync(pipeInfo).ConfigureAwait(False)
            Await System.Console.Out.FlushAsync().ConfigureAwait(False)
            _logger.LogInformation("Pipe ready, name output to stdout: {PipeInfo}", pipeInfo)

            _logger.LogDebug("Waiting for client connection on pipe: {PipeName}", _pipeName)
            Await connectionTask.ConfigureAwait(False)

            _readStream = _pipeServer
            _writeStream = _pipeServer
            _isConnected = True

            _logger.LogInformation("Client connected to named pipe")
        End Function

        ''' <summary>
        ''' Gets the full pipe path that clients should use to connect.
        ''' </summary>
        Private Function GetFullPipePath() As String
            If RuntimeInformation.IsOSPlatform(OSPlatform.Windows) Then
                Return $"\\.\pipe\{_pipeName}"
            End If

            ' On Unix, the pipe name is already the full socket path
            Return _pipeName
        End Function

        Public Async Function ReadMessageAsync(Optional cancellationToken As CancellationToken = Nothing) As Task(Of String) Implements ITransport.ReadMessageAsync
            If Not _isConnected OrElse _readStream Is Nothing Then
                Throw New InvalidOperationException("Transport is not connected")
            End If

            Try
                Dim contentLength = Await ReadContentLengthAsync(cancellationToken).ConfigureAwait(False)
                If contentLength Is Nothing Then
                    _logger.LogDebug("End of pipe stream reached")
                    Return Nothing
                End If

                Dim buffer(contentLength.Value - 1) As Byte
                Dim totalRead = 0
                While totalRead < contentLength.Value
                    Dim bytesRead = Await _readStream.ReadAsync(buffer, totalRead, contentLength.Value - totalRead, cancellationToken).ConfigureAwait(False)
                    If bytesRead = 0 Then
                        _logger.LogWarning("Unexpected end of stream while reading message body")
                        Return Nothing
                    End If

                    totalRead += bytesRead
                End While

                Dim message = Encoding.UTF8.GetString(buffer)
                _logger.LogTrace("Received message via pipe: {Message}", message)
                Return message
            Catch ex As OperationCanceledException
                Throw
            Catch ex As Exception
                _logger.LogError(ex, "Error reading message from named pipe")
                Throw
            End Try
        End Function

        Public Async Function WriteMessageAsync(message As String, Optional cancellationToken As CancellationToken = Nothing) As Task Implements ITransport.WriteMessageAsync
            If Not _isConnected OrElse _writeStream Is Nothing Then
                Throw New InvalidOperationException("Transport is not connected")
            End If

            Dim messageBytes = Encoding.UTF8.GetBytes(message)
            Dim header = $"{ContentLengthHeader}{messageBytes.Length}{vbCrLf}{vbCrLf}"
            Dim headerBytes = Encoding.UTF8.GetBytes(header)

            Await _writeLock.WaitAsync(cancellationToken).ConfigureAwait(False)
            Try
                Await _writeStream.WriteAsync(headerBytes, 0, headerBytes.Length, cancellationToken).ConfigureAwait(False)
                Await _writeStream.WriteAsync(messageBytes, 0, messageBytes.Length, cancellationToken).ConfigureAwait(False)
                Await _writeStream.FlushAsync(cancellationToken).ConfigureAwait(False)
                _logger.LogTrace("Sent message via pipe: {Message}", message)
            Finally
                _writeLock.Release()
            End Try
        End Function

        Private Async Function ReadContentLengthAsync(cancellationToken As CancellationToken) As Task(Of Integer?)
            If _readStream Is Nothing Then
                Return Nothing
            End If

            Dim headerBuilder As New StringBuilder()
            Dim buffer(0) As Byte

            While True
                Dim bytesRead = Await _readStream.ReadAsync(buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(False)
                If bytesRead = 0 Then
                    Return Nothing
                End If

                headerBuilder.Append(ChrW(buffer(0)))
                Dim headerText = headerBuilder.ToString()

                Dim headerTerminator = vbCrLf & vbCrLf
                If headerText.EndsWith(headerTerminator, StringComparison.Ordinal) Then
                    Dim lines = headerText.Split({vbCrLf}, StringSplitOptions.RemoveEmptyEntries)
                    For Each line In lines
                        If line.StartsWith(ContentLengthHeader, StringComparison.OrdinalIgnoreCase) Then
                            Dim lengthStr = line.Substring(ContentLengthHeader.Length)
                            Dim length As Integer
                            If Integer.TryParse(lengthStr, length) Then
                                Return length
                            End If

                            _logger.LogError("Invalid Content-Length value: {Value}", lengthStr)
                            Throw New InvalidOperationException($"Invalid Content-Length: {lengthStr}")
                        End If
                    Next

                    _logger.LogError("No Content-Length header found in: {Headers}", headerText)
                    Throw New InvalidOperationException("Missing Content-Length header")
                End If

                If headerBuilder.Length > 8192 Then
                    Throw New InvalidOperationException("Header too large")
                End If
            End While

            Return Nothing
        End Function

        Public Function DisposeAsync() As ValueTask Implements IAsyncDisposable.DisposeAsync
            Return New ValueTask(DisposeAsyncCore())
        End Function

        Private Async Function DisposeAsyncCore() As Task
            _isConnected = False
            _writeLock.Dispose()

            If _pipeServer IsNot Nothing Then
                Await _pipeServer.DisposeAsync().ConfigureAwait(False)
            End If

            If Not RuntimeInformation.IsOSPlatform(OSPlatform.Windows) AndAlso File.Exists(_pipeName) Then
                Try
                    File.Delete(_pipeName)
                Catch ex As Exception
                    _logger.LogWarning(ex, "Failed to delete socket file: {Path}", _pipeName)
                End Try
            End If
        End Function
    End Class

End Namespace
